Feat: improve share views - add filters

dlpzx opened the pull request:
### Feature or Bugfix
- Feature

### Detail
- Implements part 1 and 3 of the design in #644 
- It also handles the changes needed for point 6 to keep working

### Relates
- #644 

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)? -- Yes, it introduces a new GraphQL Query: listAllGroups
  - Is the input sanitized? -- N/A as no input is needed
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

noah-paige commented on backend/dataall/modules/datasets_base/db/dataset_repositories.py:
```diff
@@ -53,9 +53,12 @@ def create_dataset(
         organization = Organization.get_organization_by_uri(
             session, environment.organizationUri
         )
+        suffix_prop = config.get_property('modules.datasets.features.dataset_suffix_tag', default=None)
```
I think this was brought over from a different PR or from workshop - do not think it is needed here and may have been causing errors on my integration tests in quality gate

noah-paige commented on frontend/src/modules/Shares/components/ShareBoxList.js:
```diff
@@ -0,0 +1,489 @@
+  const [datasetGroupOptions, setDatasetGroupOptions] = useState([]);
+  const [datasets, setDatasets] = useState([]);
+  const statusOptions = [
+    'Draft',
```
Maybe for an enhancement idea - is there a better way to get the shareStatus options rather than have them hardcoded here in the frontend (...could be as a query to the backend that returns the different Enum options)

Also I see we have some hardcoded status strings in `ShareStatus` - not sure if could also extend to there?

Just an idea, but think can be taken up in a later PR

dlpzx replied:
Yes, it was in the TODOs comments. I moved it to a constant file in FE. We are starting to have the case that there is shared code between deploy, FE and BE, which leads me to think that it is time to have some globals

noah-paige commented on frontend/src/modules/Shares/components/ShareBoxList.js:
```diff
@@ -0,0 +1,489 @@
+    fetchMyGroupsOptions
+  ]);
+
+  useEffect(() => {
```
Could this `useEffect()` call `fetchItems()` function - I think the logic is the exact same for both

noah-paige commented on frontend/src/modules/Shares/components/ShareBoxListItem.js:
```diff
@@ -0,0 +1,95 @@
+            }}
+          >
+            <Typography color="textPrimary" variant="body1">
+              {`Request owner [principal]`}
```
Minor - we are using `{TEXT}` in Typography even when text is static in this file

noah-paige commented on frontend/src/modules/Shares/components/ShareBoxList.js:
```diff
@@ -0,0 +1,489 @@
+  const fetchInboxDatasetsOptions = useCallback(async () => {
+    setLoading(true);
+    const response = await client.query(
+      listDatasets(Defaults.selectListFilter)
```
To test (bc I may be missing something) - but when we call `listDatasets` would we be returning all datasets that the user is Admin or steward to + has requested via other ShareRequests (this second piece we would not want)

dlpzx replied:
Resolved in the latest commits

noah-paige commented on frontend/src/modules/Shares/components/ShareBoxList.js:
```diff
@@ -0,0 +1,489 @@
+import { ShareBoxListItem } from './ShareBoxListItem';
+import { listDatasets } from '../../Datasets/services';
+import { listAllGroups } from '../../Environments/services';
+//TODO MANAGE DEPENDENCY
```
Minor - clean up TODO

Is this in reference to dependency this view needs for Dataset to be loaded for this view to work?

dlpzx replied:
Have a  look at my last commit, By the way that the frontend modularization design was built, we will soon run into a non-modularized frontend that has everything int he shared services

noah-paige commented on frontend/src/modules/Shares/index.js:
```diff
@@ -1 +1 @@
-export { ShareInboxList } from './components';
+export { ShareBoxList } from './components';
```
Merge conflict here - need to pull latest from main to check to load `SharesModule` the way we are now managing frontend dependencies

May resolve the "Manage Dependency TODO" earlier

noah-paige commented on backend/dataall/modules/dataset_sharing/db/share_object_repositories.py:
```diff
@@ -562,6 +562,33 @@ def list_user_received_share_requests(session, username, groups, data=None):
                 )
             )
         )
+        # TODO: search functionality
```
Clean up TODO if we do not plan to add search functionality for share views and just keep to filters

noah-paige commented:
Test:
- [x] Pre-existing shares showing in ShareBox View (both Received and Sent)
- [x] Open Share Request Button works

Share Received (Inbox)
- [x] Filter By Status
- [ ] Filter by Dataset
Only showing 5 datasets, when I own 9 Datasets. I think this may be due to some pagination issues because listDataset is not picking up the filter we are trying to pass in and setting the filter to `filter = {'page': 1, 'pageSize': 5}` in `def list_datasets` (when I look at the response in the UI I see the total count is 9 but it has 2 pages)

Also showing datasets that are shared with the user but not owned by the user in the received share requests (because of listDataset()) query logic

- [ ] Filter By DatasetOwners (Group Showing 3 times)

Showing my same group 3 times since I am a part of 3 different environments and thus have 3 different EnvironmentGroup entries when calling `listAllGroups()` - think just need to add a distinct() filter here on the query or use a Set on the frontend


- [ ] Filter By RequestOwners

This works but only shows the RequestPrincipals that are for shares with principalType Group. I think this is because we are pulling the principal.SamlGroupName which will all be the group name that owns the consumption role (when of type consumption role). We may want to add some logic to pull `principalName` rather than `SamlGroupName` for consumption role types

------------------------------------------------------------------------------------------
Share Sent (Outbox)
- [x] Filter By Status
- [x] Filter by Dataset 
Unlike above, it is working when using `getShareRequestsFromMe` to determine datasets

- [x] Filter By DatasetOwners (Group Showing 3 times)
Unlike above, it is working when using `getShareRequestsFromMe` to determine datasetOwner Groups

- [ ] Filter By RequestOwners
Same issue as above with DatasetOwners (on listAllGroups)

------------------------------------------------------------------------------------------
Dataset Shares View
- [x] Pre-existing shares showing in Dataset Share Tab
- [ ] Filter by Status or Request Owner
Filters can be set but are not actually filtering the results in the Dataset Shares View

noah-paige commented:
Overall, I think the design and UI views are super and a great improvement! Left some comments on the issues I ran into, mostly minor fixes but some bigger ideas / thoughts as well

Let me know what you think @dlpzx

dlpzx commented on frontend/src/modules/Shares/components/ShareBoxList.js:
```diff
@@ -0,0 +1,489 @@
+                  />
+                </Grid>
+              )}
+              {!dataset && (
```
That was my initial design, but for some reason it showed errors and was unable to render the page

noah-paige replied:
May work if add a empty parent elements `<> </>` wrapped around the 2 `<Grid>` components but not sure if it will change frontend UI formatting, either way it is very minor and still good as is

dlpzx commented:
Solved distinct requesters owners and I am reviewing the useEffects logics

dlpzx commented on frontend/src/modules/Shares/components/ShareBoxList.js:
```diff
@@ -0,0 +1,489 @@
+import {
```
The main issue is the pagination and number of items retrieved. When we are setting the options we do not want to set the options based on the first 10 shares only.

| FILTER | request owners | roles | datasets | dataset owners |
|--------| ----------------|-----|---------|--------------------|
| INBOX (received) | ListRequests parse owners |  ListRequests parse roles | ListDatasets | ListGroups |
| OUTBOX (sent) | ListGroups | ListGroups+ListConsumptionRoles | ListRequests parse datasets | ListRequests parse datasets owners |


ListRequest
- initial execution to set filterOptions  - `pageSize=Defaults.selectListFilter`
- initial execution to fetchItems  - `pageSize=Defaults.filter`
- filter change fetchItems  - `pageSize=Defaults.filter`

ListDataset, ListGroups, ListConsumptionRoles
- initial execution to set filterOptions

noah-paige commented:
Updated Testing in AWS Deployment:
- [x] Pre-existing shares showing in ShareBox View (both Received and Sent)
- [x] Open Share Request Button works

Share Received (Inbox) - Testing Filters Show Correct Options and Work Filtering Shares
- [x] Filter By Status
- [x] Filter By RequestOwners - Showing Group Names of Requests
- [x] Filter By IAM Role - Showing IAM Roles of Requests
- [x] Filter by Dataset - Showing Datasets Owned by User’s Teams
- [x] Filter By DatasetOwners - Showing User’s Teams

------------------------------------------------------------------------------------------
Share Sent (Outbox)
- [x] Filter By Status
- [x] Filter By RequestOwners - Showing User’s Teams 
- [x] Filter By IAM Role - Showing User’s Team IAM Roles / Consumption Roles
- [x] Filter by Dataset - Showing Datasets In User’s Shares Sent
- [x] Filter By DatasetOwners - Showing Dataset Owners of Datasets in User’s Shares Sent

------------------------------------------------------------------------------------------
Dataset Shares View
- [x] Pre-existing shares showing in Dataset Share Tab
- [ ] Filter By Status, Request Owners, and Request IAM Role Names —> Showing same as inbox shares view

My one comment is that we are returning all request owners and all request IAM role names for all existing shares (not filtering for also the particular dataset we are looking into for the dataset-shares tab). 

An enhancement could be to enforce filtering for only this dataset when calling `getShareRequestsToMe` in `fetchInboxRequestOptions` to limit the options shown for this particular tab (not sure how easy this is considering the way it is designed to re-use functions across the 3 different share views in UI)

- [ ] Filters are correctly filtering Shares show on UI

The filters can be set but do not actual filter the outputted Share Objects in the dataset share tab. `listDatasetShareObjects` API needs similar filters in the backend the same way we have to `getShareRequestsFromMe` and `getShareRequestsToMe`

dlpzx commented:
Hi @noah-paige, I changed a bit how we fetch share items in the dataset view. Instead of modifying the existing getDataset, I re-used the fetchInbox methods but setting the filter to the dataset_uri and dataset_owner. I also fixed some conflicts with the filters that were not set to the initial state on tab or window changes

noah-paige commented:
Code changes look good! Think this will fix the remaining issues I was seeing with the Dataset Shares View - to test once more and report back any findings

noah-paige reviewed (approved):
I think this PR is good to go! Looks great @dlpzx - approving now
